cylc/flow/data_store_mgr.py

```diff
@@ -766,100 +839,122 @@ def update_data_structure(self, updated_nodes=None):
-
-        # Clear deltas after application and publishing
-        self.clear_deltas()
+        if self.state_update_families:
+            self.updates_pending = True
 
-    def increment_graph_elements(self):
-        """Generate and/or prune graph elements if needed.
+    def prune_data_store(self):
+        """Remove old nodes and edges by cycle point.
 
-        Use the task pool and edge source/target cycle points to find
-        new points to generate edges and/or old points to prune data-store.
+        Args:
+            point_strings (iterable):
+                Iterable of valid cycle point strings.
 
         """
-        # Gather task pool cycle points.
-        old_pool_points = self.pool_points.copy()
-        self.pool_points = set(self.schd.pool.pool)
-        # No action if pool is not yet initiated.
-        if not self.pool_points:
+        self.prune_pending = False
+        if not self.prune_flagged_nodes:
             return
-        # Increment edges:
-        # - Initially for each cycle point in the pool.
-        # - For each new cycle point thereafter.
-        # Using difference and pointwise allows for historical
-        # task insertion (in gaps).
-        new_points = self.pool_points.difference(old_pool_points)
-        if new_points:
-            for point in new_points:
-                # All family & task cycle instances are generated and
-                # populated with static data as 'ghost nodes'.
-                self.generate_graph_elements(point, point)
-            self.min_point = min(self.pool_points)
-            self.max_point = max(self.pool_points)
-        # Prune data store by cycle point where said point is:
-        # - Not in the set of pool points.
-        # - Not a source or target cycle point in the set of edges.
-        # This ensures a buffer of sources and targets in front and behind the
-        # task pool, while accommodating exceptions such as ICP dependencies.
-        # TODO: Turn nodes back to ghost if not in pool? (for suicide)
-        prune_points = set()
-        for s_point, t_points in list(self.edge_points.items()):
-            if (s_point not in self.pool_points and
-                    t_points.isdisjoint(self.pool_points)):
-                prune_points.add(str(s_point))
-                prune_points.update((str(t_p) for t_p in t_points))
-                del self.edge_points[s_point]
+
+        # Gather nodes to prune via a diff of nodes in the dependency paths
+        # of flagged boundary nodes against all other paths.
+        in_paths = [
+            v
+            for k, v in self.n_window_nodes.items()
+            if k in self.all_task_pool
+        ]
+        #    if k in self.all_task_pool
+        out_paths = [
+            v
+            for k, v in self.n_window_nodes.items()
+            if k in self.prune_flagged_nodes
+        ]
+        node_ids = set().union(*out_paths).difference(*in_paths)
```

Comment:
Would `set(*out_paths)` produce the same as `set().union(*out_paths)` without the `union` (since the `set()` produces an empty set?)?

Reply:
I thought so too.. but I think it would be a set of sets